Use JSDoc for documentation

jgraham opened the pull request:
The goal here is to address some complaints around the quality of the API documentation, and to make the docs easier to maintain going forward. To that end there are two big changes:
* The addition of sphinx-js to enable adding JSDoc strings directly into the docs.
* Conversion of the testharness and testdriver docs to use the JSDoc rather than repeating all the documentation in the markdown files as well as in the source.

There are also some incidential changes:
* The readdition of a Dockerfile for generating docs. The `wpt build-docs` command is updated to optionally use this docker image, avoiding the need to install the js components locally.
* The addition of sphinx-autoreload so that it's possible to work on the docs using a local server and have the output regenerated when the inputs change.

jgraham committed:
Use wpt build-docs to actually build the docs

This removes the previous Makefile, to simplify things by only having
a single supported entrypoint and a single programming language for
docs building.

jgraham committed:
Use source-dir relative paths in toc includes

jgraham committed:
Upgrade Sphinx to 3.x

This is in preparation for adding sphinx-js, which seems to be tested
with the 3.x branch of sphinx

jgraham committed:
Add support for building js docs with sphinx-js

This requires jsdoc to be installed using `npm install -g jsdoc`.

jgraham committed:
Add a docker image for building docs

This adds a docker container (which locally is tagged as wpt:docs) to
enable building the documentation without installing system level
dependencies. The `wpt build-docs` command is updated to allow running
the build in the docker image using `./wpt build-docs --docker`.

jgraham committed:
Add support for running a docs server with auto-reload.

This adds a `--serve` option to `wpt build-docs` which will run a
local server. sphinx-autobuild is used so that the docs are
automatically rebuilt when relevant files change.

jgraham committed:
Tests can't have a status of ERROR, so don't try to set that in assert_wrapper

jgraham committed:
Rewrite testharness API docs

This extracts more of the documentation from the JSDoc comments in the
file itself and reorganises things a little so the documentation is
easier to navigate.

jgraham committed:
Rewrite the testdriver docs.

This extracts most of the documentation from the JSDoc comments rather
than maintaining it in two places.

Due to the limitations of sphinx-js, it seems like every command needs
to be included one-by-one in the docs. However that has the advantage
that we can give them some orgnaisation rather than just providing a
flat list of commands.

jgraham committed:
Add ToC to tutorial

jgraham committed:
Use repo from environment rather than hardcoding

jgraham commented:
Rendered changes (current state):
https://jgraham.github.io/web-platform-tests/writing-tests/testharness-api.html
https://jgraham.github.io/web-platform-tests/writing-tests/testdriver.html

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+
+
+def unlink_source_dirs(created):
+    # Sort backwards in lenght to remove all files before getting to directory
```
```suggestion
    # Sort backwards in length to remove all files before getting to directory
```

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+            else:
+                if (not os.path.islink(dest) or
+                    os.path.join(os.path.dirname(dest), os.readlink(dest)) != src):
+                    # The fiel exists but it isn't a link or points at the wrong target
```
```suggestion
                    # The file exists but it isn't a link or points at the wrong target
```

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+    subprocess.check_call(["docker",
+                           "build",
+                           "--pull",
+                           "--tag", "wpt:docs",
```
```suggestion
                           "--tag", tag,
```

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+    cmd.extend(["--mount",
+                 "type=bind,source=%s,target=/app/web-platform-tests" % wpt_root])
+    if kwargs["serve"] is not None:
+        cmd.extend(["--expose", str(kwargs["serve"]),
+                    "--publish", f"{kwargs['serve']}:{kwargs['serve']}"])
```
Maybe
```suggestion
        serve = str(kwargs["serve"])
        cmd.extend(["--expose", serve, "--publish", f"{serve}:{serve}"])
```

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+    if kwargs["serve"] is not None:
+        cmd.extend(["--serve", str(kwargs["serve"])])
+    logger.debug(" ".join(cmd))
+    proc = subprocess.Popen(cmd)
+    proc.wait()
+    return proc.returncode
```
Is this just a long way of writing `subprocess.call`?

jgraham replied:
Yes :p

Ms2ger commented on docs/frontend.py:
```diff
@@ -1,8 +1,129 @@
+    return proc.returncode
+
 
+def build(_venv, **kwargs):
```
What's with the `_venv`?

jgraham replied:
This gets called as a command (i.e. `wpt build-docs` ends up here) and we pass the virtualenv as the first argument but don't actually use it.

foolip commented:
@jgraham there are conflicts to resolve here. I guess that this is ready for review apart from that though?

jgraham commented:
Yes, this is ready for review.

foolip commented on docs/docker/Dockerfile:
```diff
@@ -0,0 +1,35 @@
+  && pip install virtualenv
+
+# Pin to the same version that gecko is using, since that's well tested with sphinx-js.
+RUN npm install -g 'jsdoc@3.6.7'
```
Since this also appears in documentation.yml, can you cross-link them so they stay in sync? Or put it in a package-lock.json, if it's possible to `npm install -g` from that?

jgraham replied:
I've created a `package.json` file under `docs` and started using that.

foolip commented on docs/frontend.py:
```diff
@@ -1,8 +1,127 @@
+wpt_root = os.path.abspath(os.path.join(here, ".."))
+
+# Directories relative to the wpt root that we want to include in the docs
+# Sphinx doesn't support including files outside of docs/ so we temporarily symlink
```
Ha, wonderful!

foolip commented on docs/frontend.py:
```diff
@@ -1,8 +1,127 @@
+
+
+def unlink_source_dirs(created):
+    # Sort backwards in length to remove all files before getting to directory
```
Clever. I'm sure this doesn't work in the general case given some symlinks, but it should be reliable here.

foolip commented on docs/frontend.py:
```diff
@@ -1,8 +1,127 @@
+    cmd.extend(["--mount",
+                 "type=bind,source=%s,target=/app/web-platform-tests" % wpt_root])
+    if kwargs["serve"] is not None:
+        serve = str(kwargs["serve"])
```
Aha, it needs to be a string because it's used an argument. (Not because of the f-string.)

foolip commented on docs/requirements.txt:
```diff
@@ -1,3 +1,5 @@
+sphinx==3.5.4
```
Not a big problem, but can you group sphinx* and recommonmark separately?

foolip commented on docs/writing-tests/testharness-api.md:
```diff
@@ -24,30 +29,58 @@ Refer to the [Web Workers](#web-workers) section for details and an example on
+Execution of tests on a page is subject to a global timeout. By
+default this is 10s, but a test runner may set a timeout multiplier
+which alters the value according to the requirements of the test
+environment (e.g. to give a longer timeout for debug bulds).
```
```suggestion
environment (e.g. to give a longer timeout for debug builds).
```

foolip commented on resources/testdriver.js:
```diff
@@ -78,9 +78,17 @@
          *
-         * https://html.spec.whatwg.org/#triggered-by-user-activation
+         * See `triggered by user activation
+         * <https://html.spec.whatwg.org/#triggered-by-user-activation>`_.
```
What's the trailing underscore doing here?

foolip replied:
Hmm, there's plenty of this, so it's probably a supported syntax that does something useful.

foolip commented on resources/testdriver.js:
```diff
@@ -118,9 +126,21 @@
-         * This matches the behaviour of the {@link
-         * https://w3c.github.io/webdriver/#element-click|WebDriver
-         * Element Click command}.
+         * If ``element`` isn't inside the
```
```suggestion
         * If `element` isn't inside the
```

Right? Just below it's a single backquote.

jgraham replied:
No, the opposite. The way this works is that even though the documentation file is markdown, the JSDoc ends up in an `eval_rst` block and is interpreted as ReST. That uses the double backtick for inline literals (c.f. [the docs](https://docutils.sourceforge.io/docs/user/rst/quickref.html#inline-markup)).

We also end up using inline ReST for all the parts of JSDoc that sphinx-js doesn't support (which is quite a lot of it).

foolip commented on resources/testharness.js:
```diff
@@ -619,6 +645,20 @@
+     *
+     * Promise tests are tests which are represeted by a promise
+     * object. If the promise is fulfilled the test passes, if it's
+     * rejected with an :js:class:`AssertionError` the test fails,
```
This isn't true is it? Doesn't a promise test just fail if rejected with any value, not just `AssertionError`?

jgraham replied:
Correct, apparently I'd forgotten how this actually worked when it came to this part.

foolip commented on resources/testharness.js:
```diff
@@ -721,8 +768,21 @@
      * optional, argument the assertion description.
+     *
+     * @param {Test} test - the `Test` to use for the assertion.
+     * @param {number|string} type -  See documentation for assert_throws_dom.
```
Can `assert_throws_dom` be linked somehow?

foolip commented on resources/testharness.js:
```diff
@@ -2101,6 +2544,22 @@
+     *
+     * @example
+     * let t = async_test("Example");
+     * onload = t.step_func(e => {assert_equals(e.name, "load"; t.done())})
```
```suggestion
     * onload = t.step_func(e => {assert_equals(e.name, "load"); t.done()})
```

But I wonder if omitting `t.done()` wouldn't make for a better example, since otherwise this could use `t.step_func_done`...

jgraham replied:
I think it's hard to think of an example that's, small, complete, and doesn't benefit from `step_func_done`. Given that I'd rather do something slightly non-idomatic, but correct, than have an example that's either complex or incomplete.

foolip commented on resources/testharness.js:
```diff
@@ -1896,15 +2278,43 @@
 
+    /**
+     * Asserts if called. Used to ensure that a specific codepath is
+     * not taken e.g. that an error event isn't fired..
```
```suggestion
     * not taken e.g. that an error event isn't fired.
```

foolip commented on resources/testharness.js:
```diff
@@ -1669,20 +2042,13 @@
         }
     }
 
+    // TODO: Figure out how to document the overloads better.
```
Glad I came across this TODO. I also wondered when reading some of the `thisOrThat` parameter names, if there isn't a way to describe the valid function signatures, more like what MDN does now: https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener#syntax

But I guess that's not very JS-y and it's not too surprising if it's not supported.

jgraham replied:
So JSDoc kind of supports overloads c.f. https://github.com/jsdoc/jsdoc/issues/1017 but I couldn't get it working with sphinx-js (c.f. https://github.com/mozilla/sphinx-js/issues/152 although that's about typescript). This was definitely about the point I started to consider the merits of writing a replacement for sphinx-js ;)

foolip commented:
@jgraham I've reviewed all of this now, and went with individual comments since I started there, but I see I shouldn't have.

Anyway, a lot has moved around in this PR and it's hard to know where to pay more attention. Is there any big chunk of documentation that's entirely new that I should look much closer at?

foolip committed:
Update docs/writing-tests/testharness-api.md

foolip committed:
Update resources/testharness.js

jgraham committed:
Use package.json to specify node deps

jgraham committed:
Sort docs requirements

jgraham committed:
Update links in testharness.js

jgraham committed:
Update step_func documentation

jgraham commented:
In terms of things to look closely at:
https://jgraham.github.io/web-platform-tests/writing-tests/testharness-api.html
https://jgraham.github.io/web-platform-tests/writing-tests/testdriver.html

jgraham committed:
Fix some broken links

foolip commented on docs/Dockerfile:
```diff
@@ -29,7 +29,8 @@ RUN apt-get -qqy update \
+
+COPY ./package.json ./
+RUN npm install .
+ENV PATH=/app/node_modules/.bin:$PATH
```
This might be necessary, but maybe `npx` could be used, at least if invoked from /app/.

foolip commented on docs/package.json:
```diff
@@ -0,0 +1,11 @@
+  "repository": {},
+  "license": "BSD",
+  "devDependencies": {
+    "jsdoc": "^3.6.7"
```
Best to pin this:

```suggestion
    "jsdoc": "3.6.7"
```

foolip commented on resources/testharness.js:
```diff
@@ -2818,6 +2825,13 @@
+     *
+     * Manually set the test status to ``TIMEOUT``.
+     *
+     * Synonym for `Test.timeout <#Test.timeout>`_.
```
Suggesting a word I see used more often about APIs:

```suggestion
     * Alias for `Test.timeout <#Test.timeout>`_.
```

jgraham committed:
Run npm install in the docs directory in CI

Trying to run it in the parent seems to not work in npm 6.x

jgraham committed:
Apply suggestions from code review

Co-authored-by: Philip Jägenstedt <philip@foolip.org>

jgraham committed:
Fix documentation of Test.force_timeout

I couldn't get sphinx-js to work with @name or @function, so it seems
to be easier just to make an actual wrapper function so the docs render correctly.
